code/ship/shiphit.cpp

```diff
@@ -269,6 +269,14 @@ void do_subobj_destroyed_stuff( ship *ship_p, ship_subsys *subsys, vec3d* hitpos
 		}
 	}
 
+	// call a scripting hook for the subsystem (regardless of whether it's added to the mission log)
+	Script_system.SetHookObject("Ship", ship_objp);
+	Script_system.SetHookVar("SubsysName", 's', psub->name);
+	Script_system.SetHookVar("SubobjName", 's', psub->subobj_name);
+	Script_system.SetHookVar("SubsysIndex", 'i', subsystem_index);
```

Comment:
FYI, you can also set more complex scripting handles here. I'm not sure if all the information is there for supplying a subsystem handle here but I think that could be useful.

Reply:
Did you have something particular in mind?

Reply:
Just looked it up. You can use `ship_subsys_h` to construct a handle to a subsystem and simply use that like this:
```c++
	Script_system.SetHookVar("Subsystem", 'o', ship_subsys_h(ship_objp, subsys));
```

Reply:
Ah, brilliant.  Will do.

Reply:
That didn't exactly work.  First I got "identifier not found" and I realized I had to use this instead:

```
Script_system.SetHookVar("Subsystem", 'o', scripting::api::ship_subsys_h(ship_objp, subsys));
```

But then I got these errors:

```
1>------ Build started: Project: code, Configuration: Debug Win32 ------
1>cl : Command line warning D9025: overriding '/Zm200' with '/Zm170'
1>shiphit.cpp
1>c:\my documents\visual studio projects\visual c++\fs2open.github.com-goober5000-warpdist\code\scripting\ade_args.h(325): error C2665: 'scripting::internal::set_single_arg': none of the 5 overloads could convert all the argument types
1>c:\my documents\visual studio projects\visual c++\fs2open.github.com-goober5000-warpdist\code\scripting\ade_args.h(305): note: could be 'void scripting::internal::set_single_arg(lua_State *,char,const luacpp::LuaFunction &)'
1>c:\my documents\visual studio projects\visual c++\fs2open.github.com-goober5000-warpdist\code\scripting\ade_args.h(304): note: or       'void scripting::internal::set_single_arg(lua_State *,char,luacpp::LuaFunction *)'
1>c:\my documents\visual studio projects\visual c++\fs2open.github.com-goober5000-warpdist\code\scripting\ade_args.h(302): note: or       'void scripting::internal::set_single_arg(lua_State *,char,const luacpp::LuaTable &)'
1>c:\my documents\visual studio projects\visual c++\fs2open.github.com-goober5000-warpdist\code\scripting\ade_args.h(301): note: or       'void scripting::internal::set_single_arg(lua_State *,char,luacpp::LuaTable *)'
1>c:\my documents\visual studio projects\visual c++\fs2open.github.com-goober5000-warpdist\code\scripting\ade_args.h(293): note: or       'void scripting::internal::set_single_arg(lua_State *,char,const char *)'
1>c:\my documents\visual studio projects\visual c++\fs2open.github.com-goober5000-warpdist\code\scripting\ade_args.h(325): note: while trying to match the argument list '(lua_State *, const char, _Ty)'
1>        with
1>        [
1>            _Ty=scripting::api::ship_subsys_h
1>        ]
1>c:\my documents\visual studio projects\visual c++\fs2open.github.com-goober5000-warpdist\code\scripting\ade_args.h(351): note: see reference to function template instantiation 'void scripting::internal::set_args_actual<_Ty,>(lua_State *,scripting::internal::set_args_state &,const char *,T &&)' being compiled
1>        with
1>        [
1>            _Ty=scripting::api::ship_subsys_h,
1>            T=scripting::api::ship_subsys_h
1>        ]
1>c:\my documents\visual studio projects\visual c++\fs2open.github.com-goober5000-warpdist\code\scripting\scripting.h(259): note: see reference to function template instantiation 'int scripting::ade_set_args<_Ty>(lua_State *,const char *,_Ty &&)' being compiled
1>        with
1>        [
1>            _Ty=scripting::api::ship_subsys_h
1>        ]
1>c:\my documents\visual studio projects\visual c++\fs2open.github.com-goober5000-warpdist\code\ship\shiphit.cpp(276): note: see reference to function template instantiation 'void script_state::SetHookVar<scripting::api::ship_subsys_h>(const char *,char,T &&)' being compiled
1>        with
1>        [
1>            T=scripting::api::ship_subsys_h
1>        ]
1>Done building project "code.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```

Reply:
Ah, sorry. You need to use `l_Subsystem.Set(<ship_subsys_h constructor>)`.

Reply:
So this, then?

```
	Script_system.SetHookVar("Subsystem", 'o', scripting::api::l_Subsystem.Set(scripting::api::ship_subsys_h(ship_objp, subsys)));
```

I am a little bit skeptical because it looks like neither the `l_Xxxxx` classes nor the `xxx_xxx_h` structs are used outside the scripting API implementation.

Reply:
The `l_<something>` values are the "bridge" between C++ and Lua. They have some special logic (along with `SetHookVar` that transfer the value you specify as the parameter into Lua so that they can be used as scripts.

The reason you do not find `scripting::api::l_Something.Set` is that most of the Lua API code doesn't need the namespace prefix. Also, the system which allows to declare the API types in headers is relatively new so it isn't used that widely yet but I can assure you that the way you have written it should work.